utils/checks.stanza

```diff
@@ -113,8 +133,207 @@ public pcb-check rated-temperature (component:JITXObject):
+           description =          "Check io pin properties"
+           category =             "pin-check"
+           subcheck-description = "Check that no-connect pads are not connected",
+           pass-message =         "Pin %_ is marked no-connect and is correctly not connected" % [ref(p)],
+           fail-message =         "Pin %_ is marked no-connect and is incorrectly connected to other pins" % [ref(p)],
+           locators =             [p])
+
+    for out-p in connected-io-pins do :
+      val driver = 
+        if has-property?(out-p.digital-io) : 
+          driver(property(out-p.digital-io))
+        else : 
+          driver(property(out-p.digital-output))
+      match(driver) : 
+        (o:CMOSOutput|TTLOutput) : 
+          cmos-and-ttl-checks(out-p, connected-io-pins)
+        (o:OpenCollector) : 
+          open-collector-checks(out-p, o, all-connected-pins)
+ 
+; Checks any IO pins marked CMOSOutput or TLLOutput
+defn cmos-and-ttl-checks (out-p:JITXObject,
+                          connected-io-pins) : 
+  for in-p in filter({_ != out-p}, connected-io-pins) do :
+    ; run the checks
+    check-single-push-pull-levels(out-p, in-p)
+
+; Checks that run on a pin marked OpenCollector
+; - io-pin: the pin we are checking
+; - driver: the driver (value of driver(property(io-pin.digital-io)))
+; - all-connected-pins: the pins this pin is connected to
+;
+defn open-collector-checks (io-pin:JITXObject, 
+                            driver:OpenCollector,
+                            all-connected-pins:Tuple<JITXObject>) : 
+  ; First, we search for the pull-up-resistors that might be connected to this pin
+  val pull-up-resistors = to-tuple $
+    for p in all-connected-pins seq? : 
+      val instance = containing-instance(p) as Instance
+      val connected-to-rail? = 
+        for p in pins(instance) any? : 
+          has-property?(p.rail-voltage)
+      if has-property?(instance.resistance) and connected-to-rail? : 
+        One(instance) 
+      else : 
+        None()
+
+  ; Check that there exists at least one pull-up-resistor
+  #CHECK(
+    condition =            not empty?(pull-up-resistors),
+    name =                 "IO pin checks"
+    description =          "Check io pin properties"
+    category =             "pin-check"
+    subcheck-description = "Check that io pins marked OpenCollector have pullup resistors",
+    pass-message =         "Pin %_ is marked OpenCollector and has pull-up resistor(s)" % [ref(io-pin)],
+    fail-message =         "Pin %_ is marked OpenCollector and is missing pull-up resistor(s)" % [ref(io-pin)],
+    locators =             [io-pin])
+
+  ; Next, we check that all the pull-up resistors are connected to the same rail
+  val same-rail-voltage? = all-equal? $ 
+    for resistor in pull-up-resistors seq :
+      for p in pins(resistor) first! : 
+        property?(p.rail-voltage)
+    
+  #CHECK(
+    condition =            same-rail-voltage?,
```

Comment:
Can we add or replace this check with one that checks that all resistors are connected to the same net (vdd-like)? That is probably more germane to the check and allows for switchable rail nets. In other words, the (vdd) net for all resistors should be the same (similar to the rail-voltage property).

Reply:
Sure, I'm looking into this. I'm not sure if we support net-level introspection for connected pins yet, but this is a good use case.

Reply:
If we can't check the net itself, then can we check the pins on that net (vdd) for a PowerPin property?

Reply:
Added!